### PR TITLE
`Integrated code lifecycle`: Add editing of auxiliary repositories in the online code editor for instructors

### PR DESCRIPTION
**Pull request by SimonEntholzer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Instructors can edit Template, Assignment, Solution and Test repositories in Artemis, by using the "edit in editor" view.
This is not yet possible for auxiliary repositories. 

### Description

This PR adds support for auxiliary repositories in the edit in editor view.
Also fixes a small UI bug, where the text on the drop-down button was not readable, when not hovering.

### Steps for Testing

1. Create a programming exercise with one or more auxiliary repositories
2. In the exercise details page, open the "edit in editor" view
3. In the drop-down, select the previously created auxiliary repositories.
4. Add/Delete/Change files and folders, and submit.
5. Go back to the details view, clone the auxiliary repositories and check that the created content is there.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| code-editor-instructor-and-editor-container.component.ts | 100% | ✅ |
| code-editor-instructor-base-container.component.ts | 83.33% | ✅ ❌ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| ProgrammingExerciseRepository.java | 85% | ✅ ❌ |

### Screenshots

image
image

small Bugfix:
image
vs. 
image

## Release Notes

- **New Features**
  - Enhanced repository selection in the code editor to include auxiliary repositories.
  - New route for accessing auxiliary repositories with appropriate access control.

- **Bug Fixes**
  - Improved data retrieval for programming exercises, ensuring auxiliary repositories are included.

- **Style**
  - Updated button styles for better visibility of selected auxiliary repositories.
  - Introduced new CSS rules for button styling in the code editor.

- **Chores**
  - Refined component structure to better manage auxiliary repository selections.
  - Updated tests to validate new functionality related to auxiliary repositories.

### CODE REVIEW

**Review by sachmii — APPROVED** [review]
Tested on TS4, after adding files and folders to the auxiliary repository and cloning it, they also appeared locally. Worked just fine.

Bildschirmfoto 2024-10-24 um 22 41 35
Bildschirmfoto 2024-10-24 um 22 41 41

File: src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseRepository.java [hunk 1]
```diff
@@ -776,8 +776,10 @@ default ProgrammingExercise findByIdWithTemplateAndSolutionParticipationLatestRe
                 programmingExerciseId);
         ProgrammingExercise programmingExerciseWithSolution = getValueElseThrow(findWithSolutionParticipationLatestResultFeedbackTestCasesById(programmingExerciseId),
                 programmingExerciseId);
+        ProgrammingExercise programmingExerciseWithAuxiliaryRepositories = findByIdWithAuxiliaryRepositoriesElseThrow(programmingExerciseId);
 
         programmingExerciseWithTemplate.setSolutionParticipation(programmingExerciseWithSolution.getSolutionParticipation());
+        programmingExerciseWithTemplate.setAuxiliaryRepositories(programmingExerciseWithAuxiliaryRepositories.getAuxiliaryRepositories());
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/cit/aet/artemis/programming/repository/ProgrammingExerciseRepository.java:782]: Can you adapt the method name to reflect the new values as well?

File: src/main/webapp/app/exercises/programming/manage/code-editor/code-editor-instructor-and-editor-container.component.html [hunk 2]
```diff
@@ -79,6 +81,20 @@
...
                         jhiTranslate="artemisApp.editor.repoSelect.testRepo"
                     ></button>
+                    @for (auxiliaryRepository of exercise.auxiliaryRepositories; track exercise.auxiliaryRepositories; let i = $index) {
+                        @if (auxiliaryRepository.id) {
+                            <button
+                                [disabled]="!exercise"
+                                (click)="selectAuxiliaryRepository(auxiliaryRepository.id)"
+                                ngbDropdownItem
+                                [style.background-color]="
+                                    selectedRepository === REPOSITORY.AUXILIARY && selectedRepositoryId === auxiliaryRepository.id ? '#3e8acc' : 'transparent'
```
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/manage/code-editor/code-editor-instructor-and-editor-container.component.html:91]: Our documentation:
> Do not hard code any color values in your component stylesheets, templates or TypeScript files!

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by HawKhiem — COMMENTED** [review]
I'm sorry but how can I create an auxiliary repository? I can't seem to find anything in the edit view and then edit in editor view.

image
image

**Comment by SimonEntholzer** [comment]
> I'm sorry but how can I create an auxiliary repository? I can't seem to find anything in the edit view and then edit in editor view.

you can only create them when initially creating a programming exercise:
image

File: src/main/webapp/app/exercises/programming/manage/code-editor/code-editor-instructor-and-editor-container.component.html (hunk 2 above)
Comment (SimonEntholzer) [reply src/main/webapp/app/exercises/programming/manage/code-editor/code-editor-instructor-and-editor-container.component.html:91]: I replaced them with --link-color :+1:

**Review by HawKhiem — DISMISSED** [review]
Tested on TS3. Works as described!

Screenshot 2024-10-25 184722
Screenshot 2024-10-25 184744
Screenshot 2024-10-25 184801
Screenshot 2024-10-25 184810

**Review by sachmii — APPROVED** [review]
Tested on TS4, still works fine. Re-approve

**Review by sarpsahinalp — APPROVED** [review]
Code lgtm

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by sachmii — APPROVED** [review]

**Review by HawKhiem — APPROVED** [review]
LGTM

**Review by JohannesStoehr — APPROVED** [review]
New tests also look good

**Pull request merged**